src/test/hardfork/archive_migration_tests/steps.ml

```diff
@@ -0,0 +1,599 @@
+    in
+    let migrated_pool =
+      Caqti_async.connect_pool ~max_size:128 migrated_archive_uri
+    in
+    match (mainnet_pool, migrated_pool) with
+    | Error _e, _ | _, Error _e ->
+        failwith "Failed to create Caqti pools for Postgresql"
+    | Ok mainnet_pool, Ok migrated_pool ->
+        let query_mainnet_db = Mina_caqti.query mainnet_pool in
+        let query_migrated_db = Mina_caqti.query migrated_pool in
+        let compare_hashes ~fetch_data_sql ~find_element_sql name =
+          let%bind expected_hashes = query_mainnet_db ~f:fetch_data_sql in
+          Deferred.List.iter expected_hashes ~f:(fun hash ->
+              let%bind element_id = find_element_sql hash in
+              if element_id |> Option.is_none then
+                failwithf "Cannot find %s hash ('%s') in migrated database" name
+                  hash ()
+              else Deferred.unit )
+        in
+        let%bind _ =
+          compare_hashes
+            ~fetch_data_sql:(fun db ->
+              Sql.Mainnet.user_commands_hashes db end_global_slot )
+            ~find_element_sql:(fun hash ->
+              query_migrated_db ~f:(fun db ->
+                  Sql.Berkeley.find_user_command_id_by_hash db hash ) )
+            "user_commands"
+        in
+        let%bind _ =
+          compare_hashes
+            ~fetch_data_sql:(fun db ->
+              Sql.Mainnet.internal_commands_hashes db end_global_slot )
+            ~find_element_sql:(fun hash ->
+              query_migrated_db ~f:(fun db ->
+                  Sql.Berkeley.find_internal_command_id_by_hash db hash ) )
+            "internal_commands"
+        in
+        let%bind _ =
+          match full_migration with
+          | true ->
+              assert_migrated_db_contains_only_canonical_blocks
+                query_migrated_db
+          | false ->
+              assert_migrated_db_contains_orphaned_and_pending_blocks
+                query_migrated_db
+        in
+        let%bind _ =
+          compare_hashes
+            ~fetch_data_sql:(fun db ->
+              match full_migration with
+              | true ->
+                  Sql.Mainnet.block_hashes db end_global_slot
+              | false ->
+                  Sql.Mainnet.block_hashes_no_orphaned db end_global_slot )
+            ~find_element_sql:(fun hash ->
+              query_migrated_db ~f:(fun db ->
+                  Sql.Berkeley.find_block_by_state_hash db hash ) )
+            "block_state_hashes"
+        in
+        let%bind _ =
+          compare_hashes
+            ~fetch_data_sql:(fun db ->
+              match full_migration with
+              | true ->
+                  Sql.Mainnet.block_parent_hashes db end_global_slot
+              | false ->
+                  Sql.Mainnet.block_parent_hashes_no_orphaned db end_global_slot
+              )
+            ~find_element_sql:(fun hash ->
+              query_migrated_db ~f:(fun db ->
+                  Sql.Berkeley.find_block_by_parent_hash db hash ) )
```

Comment:
This doesn't quite make sense. Because `compare_hashes` just collects all the data using `fetch_data_sql` and then verifies that `find_element_sql` would returns. It didn't check for that the same block has the same parent.
What you have here is something like:
```
let expected_hashes = Mina_caqti.query ~f:(fun db -> Sql.Mainnet.block_parent_hashes db end_global_slot) in
Deferred.List.iter expected_hashes ~f:(fun hash ->
  let%bind element_id = Mina_caqti.query ~f:(fun db -> Sql.Berkeley.find_block_by_parent_hash db hash) in
  if element_id |> Option.is_none then
    failwithf "Cannot find %s hash ('%s') in migrated database" name hash ()
  else Deferred.unit )
```
So it's just first get the list of blocks that have parent hashes before `end_global_slot` and then search the migrated database for blocks whose parent_hash is one of those.
What you could do is that returns both the `state_hash` and `parent_hash` from mainnet table and then search for `state_hash` and `parent_hash` in the migrated table.

Btw the way, `Sql.Mainet.block_parent_hashes` and `Sql.Mainnet.block_hashes` would returns the same result, since every block has a parent_hash (even the genesis block), so you might want to combine this comparison with "block_state_hashes"

Reply:
Good idea. i will fix it